Assignment-4

smxlikith opened the pull request:
This pull request introduces a new C program for sonar image processing, including matrix generation, rotation, and smoothing filter operations. The changes add both documentation and a complete implementation of the required features.

**New Features and Implementation:**

*Matrix Operations and Algorithms:*
- Implemented random N×N matrix generation with intensity values between 0 and 255 in `generateMatrix` and `main` functions.
- Added in-place 90° clockwise matrix rotation using pointer manipulation in `rotateMatrix`.
- Developed a 3×3 smoothing filter, replacing each cell with the average of itself and its valid neighbors via `smoothing3x3` and `applySmoothing`.

*Program Structure and Usability:*
- Included matrix display and memory cleanup functions (`display`, `cleanup`) for improved usability and resource management.
- Added a comprehensive README (`README.md`) detailing features, usage instructions, and compilation steps for the new sonar image processing program.

smxlikith committed:
Completed Assignment-4

smxlikith committed:
in-place calculation of smoothing using recursion

smxlikith committed:
Added safety checks for DMA

smxlikith committed:
Added a Non-recursive Solution

tejas0123 commented on Assignment-4/sonarImageProcessing.c:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        for (int j = 0; j < size; j++) {
+            matrix[i][j] = rand() % (MAX_VAL + 1);
```
Problem statement mentions not to use indexing anywhere. Only pointer artihmetic is allowed.

smxlikith replied:
I just skimmed through the document and overlooked it. I'll fix that in the next commit.

tejas0123 commented on Assignment-4/sonarImageProcessing.c:
```diff
@@ -0,0 +1,139 @@
+    }
+    for (int i = 0; i < size; i++) {
+        for (int j = 0; j < size / 2; j++) {
+            swap(&matrix[i][j], &matrix[i][size - j - 1]);
```
Pointer arithmetic should be used. That is the essence of the problem. Do not use indexing anywhere while referring to the matrix elements

smxlikith replied:
I will be fixing that in the next commit.

tejas0123 commented on Assignment-4/sonarImageProcessing.c:
```diff
@@ -0,0 +1,139 @@
+    }
+}
+
+int smoothing3x3(int i, int j, int size, int** matrix) {
```
You are essentially calculating the average of all neighbours. But the name does not indicate that. Name should be a verb

smxlikith replied:
I’ll be renaming the function in the latest commit to be an action.

tejas0123 commented on Assignment-4/sonarImageProcessing.c:
```diff
@@ -0,0 +1,139 @@
+    matrix[i][j] = currentSmoothedValue;
+}
+
+void applySmoothingNonRecursive(int size, int** matrix) {
```
Why is this function here? It is not being used

smxlikith replied:
I figured out how to do it non-recursively, so I’ve added that solution here as well.

tejas0123 commented on Assignment-4/sonarImageProcessing.c:
```diff
@@ -0,0 +1,139 @@
+
+    // printf("smoothing the matrix Non-recursive:\n");
+    // applySmoothingNonRecursive(size, matrix);
+    // display(size, duplicate);
```
Remove unnecessary comments

smxlikith replied:
These lines were for the non-recursive solution, but they were commented out because both the non-recursive and recursive solutions alter the matrix in place, and they cannot be run on the same matrix.

tejas0123 commented on Assignment-4/sonarImageProcessing.c:
```diff
@@ -0,0 +1,139 @@
+
+    printf("rotating the matrix\n");
+    rotateMatrix(size, matrix);
+    display(size, matrix);
```
displayMatrix

smxlikith replied:
I'll be renaming this function in the next commit.

smxlikith committed:
Update the Code:
- Fixed the Matrix references to use pointer arithmetics.
- Changed the name of the function display to displayMatrix.
- Changed the name of the function smoothing3x3 to calculate3x3NeighborsAverage.

tejas0123 reviewed (approved):
Approved. Good job finding two approaches. But remove the unused function and the commented out code and then merge

smxlikith committed:
Updated the code.
